tests/test_optuna_transformers.py

```diff
@@ -0,0 +1,168 @@
+# Copyright (c) 2021 Timothy Wolff-Piggott
+# This software is distributed under the terms of the MIT license
+# which is available at https://opensource.org/licenses/MIT
+
+import tempfile
+from pathlib import Path
+from typing import Dict
+
+import pytest
+import sklearn.model_selection
+import torch
+from mlflow.tracking import MlflowClient
+from transformers import (
+    ElectraConfig,
+    ElectraForSequenceClassification,
+    ElectraTokenizer,
+    EvalPrediction,
+    Trainer,
+    TrainingArguments,
+)
+
+import optuna
+from hpoflow.optuna_mlflow import OptunaMLflow
+from hpoflow.optuna_transformers import OMLflowCallback
+
+# global parameters for training
+MAX_STEPS = 4
+# an eval step is only logged if it coincides with a logging step
+# logging steps should divide eval steps otherwise some evals are lost
+LOGGING_STEPS = EVAL_STEPS = 2
+assert (
+    EVAL_STEPS % LOGGING_STEPS == 0
+), f"Logging steps {LOGGING_STEPS} should divide eval steps {EVAL_STEPS}"
+N_EVAL_STEPS = MAX_STEPS // EVAL_STEPS
+LAST_EVAL_STEP = N_EVAL_STEPS * EVAL_STEPS
+N_TRIALS = 2
+
+
+@pytest.fixture(scope="module")
+def config():
+    return ElectraConfig()
+
+
+@pytest.fixture(scope="module")
+def model(config):
+    return ElectraForSequenceClassification(config)
+
+
+@pytest.fixture(scope="module")
+def output_path():
+    tempdir = tempfile.mkdtemp()
+    return Path(tempdir)
+
+
+@pytest.fixture(scope="module")
+def tracking_file_name(output_path):
+    return f"file:{output_path.as_posix()}"
+
+
+def test_omlflow_callback(config, output_path, model, tracking_file_name):
+    study_name = "test_omlflow_callback"
+    study = optuna.create_study(study_name=study_name)
+    train_func = _train_func_factory(config, output_path, model, tracking_file_name)
+    study.optimize(train_func, n_trials=N_TRIALS)
+
+    mlflow_client = MlflowClient(tracking_file_name)
+
+    # test that there is 1 MLflow experiment, corresponding to the Optuna trial
+    experiments = mlflow_client.list_experiments()
+    assert len(experiments) == 1
+    experiment = experiments[0]
+    assert experiment.name == study_name
```

Comment:
Should we add `assert experiment is not None`?

Reply:
In the scenario where the experiment is not persisted correctly, `experiments` will just be an empty list and will fail the assert on line 70.

I can't think of a case where we expect the MLflow client to return a `NoneType` experiment; do you know of one?

Reply:
> do you know of one

no :-)